### DISCUSSION TITLE
How to compile openssl to load fips.dll (Also openssl.cnf) dynamically from PATH variable instead of compile time hardcoded path

### DISCUSSION THREAD

**Discussion by reguripradeep** [openssl/openssl#23479] [Q&A]
i compiled openssl 3.2.0 in one my developer machine with some folder structure like "c:\upgrades\release3\openssl" and configured prefix as "c:\upgrades\release3\openssl\Release" so that i can get all the binaries and libraries in Release folder to copy and is it on any machine.
Now, i want to use this openssl in one of my test machine where i copied the binaries to "c:\prod1\app1\openssl". The issue is, when configured only fips provider, the openssl.exe (or other dlls like lib-ssl) is trying to find fips.dll in in "c:\upgrades\release3\openssl" which is compiled time hardcoded path instead of either PATH variable or LD_LIBRARY_PATH.  one option is to use OPENSSL_MODULES environment variable. this works fine when used openssl as an individual application but this is not working when i use this openssl in another app .
for example, Apache httpd server's mod_ssl. when i start httpd server, mod_ssl is trying to load fips.dll from "c:\upgrades\release3\openssl" and it is failing as the folder structure does not exist.
this compiled time path may not exist on the actual machine so is there any other method to solve this issue?
**[Update] :** i am facing similar issue with openssl.cnf path which is again hardcoded during the compilation. i know we can use OPENSSL_CONF environment variable but this will not work with those applications which are loading libcrypto/libssl dynamically or embedded with in the application.
My question is why can't we have it in the same location from where the executable or libraries are loaded or why can't we build the binaries or libraries such that it automatically look into the same folder?

**Comment by levitte** [comment]
For cases like this, I'd start with looking what possibilities the application in question offers.

For example, there is the Apache module [mod_env](https://httpd.apache.org/docs/current/mod/mod_env.html), which allows this configuration:

    SenEnv OPENSSL_MODULES /what/ever/you/want

**Reply by reguripradeep** [reply]
already tried and did not work for me. i am not sure if  Apache is passing that variable to openssl crypto/ssl library.

**Reply by levitte** [reply]
Apache should pass those envvars, that's the whole idea with mod_env!

However, libcrypto gets environment variables with secure_getenv, so if you're setup causes what that function considers insecure (see the man page), that's a problem...

**Reply by reguripradeep** [reply]
For now , i used "module" key in fipsmodule.cnf and gave the path of fips.dll. looks like it is working. will update here once confirmed
